src/main/scala/io/pravega/connectors/spark/NonTransactionPravegaWriter.scala

```diff
@@ -0,0 +1,221 @@
+                                       scopeName: String,
+                                       streamName: String,
+                                       clientConfig: ClientConfig,
+                                       schema: StructType)
+  extends DataWriterFactory[InternalRow] with Logging {
+
+  override def createDataWriter(
+                                 partitionId: Int,
+                                 taskId: Long,
+                                 epochId: Long): DataWriter[InternalRow] = {
+    new NonTransactionPravegaDataWriter(scopeName, streamName, clientConfig, schema)
+  }
+}
+
+/**
+  * A [[DataWriter]] for Pravega writing. One data writer will be created in each partition to
+  * process incoming rows.
+  *
+  * @param inputSchema              The attributes in the input data.
+  */
+class NonTransactionPravegaDataWriter(
+                               scopeName: String,
+                               streamName: String,
+                               clientConfig: ClientConfig,
+                               inputSchema: StructType)
+  extends DataWriter[InternalRow] with Logging {
+
+  val pendingWriteCounts = new AtomicInteger(0)
+
+  private val projection = createProjection
+
+  private val clientFactory = EventStreamClientFactory.withScope(scopeName, clientConfig)
+  private val writer: EventStreamWriter[ByteBuffer] = clientFactory.createEventWriter(
+    streamName,
+    new ByteBufferSerializer,
+    EventWriterConfig
+      .builder
+      .build)
+
+
+  override def write(row: InternalRow): Unit = {
+    val projectedRow = projection(row)
+    val event = projectedRow.getBinary(1)
+    val eventToLog = if (log.isDebugEnabled) {
+      val s = new String(event, StandardCharsets.UTF_8)
+      val maxLength = 100
+      if (s.length > maxLength) s.substring(0, maxLength) + "..." else s
+    } else null
+
+    log.info(s"write: begin")
+
+    val haveRoutingKey = !projectedRow.isNullAt(0)
+    if (haveRoutingKey) {
+      val routingKey = projectedRow.getUTF8String(0).toString
+      log.debug(s"write: routingKey=${routingKey}, event=${eventToLog}")
+      writer.writeEvent(routingKey, ByteBuffer.wrap(event))
+    } else {
+      log.debug(s"write: event=${eventToLog}")
+
+      val f = toScala(writer.writeEvent(ByteBuffer.wrap(event)))
+      f
+        .map(println)
+        .recover { case e =>
+          e.printStackTrace
+          "recovered"
+        }.map(println)
+    }
+    log.debug(s"write: end")
+  }
+
+  override def commit(): WriterCommitMessage = {
+    log.debug(s"commit: begin")
+    try {
+      writer.flush()
+      NonTransactionPravegaWriterCommitMessage(null)
+    } finally {
+      close()
+      log.debug(s"commit: end")
+    }
+  }
+
+  override def abort(): Unit = {
+    log.debug("abort:BEGIN")
+      close()
+    log.debug("abort:END")
+  }
+
+  private def close(): Unit = {
+    log.debug("close: BEGIN")
+    try {
+      writer.flush()
```

Comment:
If flush fails, then the writer will not get closed. This can cause a resource leak. The easiest way to fix this is to remove writer.flush(). writer.close will flush if needed.